refactor(semantic): remove `AstNodes` `program` field

overlookmotel opened the pull request:
`Program` is always the first node in `AstNodes`, so no need to store a duplicate `&Program` reference in `AstNodes`.

I *think* this is an improvement. It does make `AstNodes::program` marginally more expensive, because it now has 2 checks:

1. Check `AstNodes::nodes` is not empty. and
2. Check that first AST node is `Program`.

But both checks are branches which are never taken, so pretty cheap, and both checks share a single panic. So overall, I think it's probably better than `AstNodes` being larger than it otherwise needs to be.

camc314 commented:
### Merge activity

* **Jul 25, 12:40 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(semantic): remove `AstNodes` `program` field (#12516)

`Program` is always the first node in `AstNodes`, so no need to store a duplicate `&Program` reference in `AstNodes`.

I *think* this is an improvement. It does make `AstNodes::program` marginally more expensive, because it now has 2 checks:

1. Check `AstNodes::nodes` is not empty. and
2. Check that first AST node is `Program`.

But both checks are branches which are never taken, so pretty cheap, and both checks share a single panic. So overall, I think it's probably better than `AstNodes` being larger than it otherwise needs to be.
